vhdl_lang/src/analysis/tests/subprogram_instance.rs

```diff
@@ -0,0 +1,521 @@
+use crate::analysis::tests::{check_no_diagnostics, LibraryBuilder};
+use crate::syntax::test::check_diagnostics;
+use crate::Diagnostic;
+
+#[test]
+pub fn cannot_instantiate_procedure_that_does_not_exist() {
+    let mut builder = LibraryBuilder::new();
+    let code = builder.in_declarative_region(
+        "\
+procedure proc
+    generic ( x: natural := 1 ) is
+begin
+end proc;
+
+procedure proc is new foo;
+    ",
+    );
+
+    let diagnostics = builder.analyze();
+    assert_eq!(
+        diagnostics,
+        vec![Diagnostic::error(
+            code.s1("foo").pos(),
+            "No declaration of 'foo'"
+        )]
+    );
+}
+
+#[test]
+pub fn instantiate_wrong_type() {
+    let mut builder = LibraryBuilder::new();
+    let code = builder.in_declarative_region(
+        "\
+signal x : bit;
+
+function proc is new x;
+    ",
+    );
+
+    let diagnostics = builder.analyze();
+    assert_eq!(
+        diagnostics,
+        vec![Diagnostic::error(
+            code.s1("new x").s1("x"),
+            "signal 'x' does not denote an uninstantiated subprogram"
+        )]
+    )
+}
+
+#[test]
+pub fn ambiguous_multiple_uninstantiated_subprograms() {
+    let mut builder = LibraryBuilder::new();
+    let code = builder.in_declarative_region(
+        "\
+procedure foo
+    generic (type T)
+    parameter (x : bit)
+is begin
+end foo;
+
+procedure foo
+    generic (type T)
+    parameter (x : bit; y: bit)
+is begin
+end foo;
+
+procedure proc is new foo;
+    ",
+    );
+
+    let diagnostics = builder.analyze();
+    check_diagnostics(
+        diagnostics,
+        vec![Diagnostic::error(
+            code.s1("new foo").s1("foo"),
+            "Ambiguous instantiation of 'foo'",
+        )
+        .related(code.s("foo", 1), "Might be procedure foo[BIT]")
+        .related(code.s("foo", 3), "Might be procedure foo[BIT, BIT]")],
+    )
+}
+
+#[test]
+pub fn by_signature_resolved_multiple_uninstantiated_subprograms() {
+    let mut builder = LibraryBuilder::new();
+    builder.in_declarative_region(
+        "\
+procedure foo
+    generic (type T)
+    parameter (x : bit)
+is begin
+end foo;
+
+procedure foo
+    generic (type T)
+    parameter (x : bit; y: bit)
+is begin
+end foo;
+
+procedure proc is new foo [bit];
+procedure proc2 is new foo [bit, bit];
+    ",
+    );
+
+    let diagnostics = builder.analyze();
+    check_no_diagnostics(&diagnostics);
```

Comment:
Doesnt the instantiation need to provide the type T?

Reply:
Fixed.
The "better" test where the parameters are generic types and one can distinguish by signature is yet another edge case I didn't address in this PR. I have added another ignored test though where the desired behavior is shown.